src/extensions/idle.rs

```diff
@@ -100,37 +140,65 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
     /// Internal helper that doesn't consume self.
     ///
     /// This is necessary so that we can keep using the inner `Session` in `wait_keepalive`.
-    fn wait_inner(&mut self, reconnect: bool) -> Result<WaitOutcome> {
+    fn wait_inner<F>(&mut self, reconnect: bool, mut callback: F) -> Result<WaitOutcome>
+    where
+        F: FnMut(UnsolicitedResponse) -> CallbackAction,
+    {
         let mut v = Vec::new();
-        loop {
-            let result = match self.session.readline(&mut v).map(|_| ()) {
+        let result = loop {
+            let rest = match self.session.readline(&mut v) {
                 Err(Error::Io(ref e))
                     if e.kind() == io::ErrorKind::TimedOut
                         || e.kind() == io::ErrorKind::WouldBlock =>
                 {
-                    if reconnect {
-                        self.terminate()?;
-                        self.init()?;
-                        return self.wait_inner(reconnect);
+                    break Ok(WaitOutcome::TimedOut);
+                }
+                Ok(_len) => {
+                    //  Handle Dovecot's imap_idle_notify_interval message
+                    if v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
+                        v.clear();
+                        continue;
+                    }
+                    match parse_idle(&v) {
+                        (_rest, Some(Err(r))) => break Err(r),
+                        (rest, Some(Ok(response))) => {
+                            if let CallbackAction::Stop = callback(response) {
+                                break Ok(WaitOutcome::MailboxChanged);
+                            }
+                            rest
+                        }
+                        (rest, None) => rest,
                     }
-                    Ok(WaitOutcome::TimedOut)
                 }
-                Ok(()) => Ok(WaitOutcome::MailboxChanged),
-                Err(r) => Err(r),
-            }?;
+                Err(r) => break Err(r),
+            };
 
-            // Handle Dovecot's imap_idle_notify_interval message
-            if v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
+            // Update remaining data with unparsed data if needed.
+            if rest.is_empty() {
                 v.clear();
-            } else {
-                break Ok(result);
+            } else if rest.len() != v.len() {
```

Comment:
This clause confuses me. How would we get here?

Reply:
In practice, we shouldn't. In theory, it is possible that the response parser does not consume the entire input and there are leftovers, so this is parsing the hypothetical leftovers.

The only reason we shouldn't get here is because we are testing on every `session.readline() ` if this is something we can parse or not. If not, we update the buffer, read another line, and try again. So as soon as the thing parses then we _should_ parse the entire input and will hit the `rest.is_empty()` clause and just clear the input vector. But if in the future we make some change or have some weird situation where the input parser does _not_ actually clear the entire input buffer, then this clause will correctly update the input.

Basically, this is effectively exhaustively matching what `rest` could be.

Reply:
Given that IMAP command responses must be terminated at newlines anyway, I think I'm comfortable saying that this simply shouldn't happen. Maybe add in a `debug_assert!(rest.is_empty())`? Also, I think I'd keep this just `else` — the `!= v.len()` bit should always be true.

Reply:
Hm. The default case here is if `v` does not parse as a complete command (multi-line response), so `rest` will not be empty and will be equal to `v` (nothing parsed). In this case we want to just loop again.

So we could just simplify the entire thing by replacing the if / else clause with this:
```
v.drain(0..(v.len() - rest.len()));
```

- If nothing parsed, then `rest == v`, and we drain `0..0`, so nothing.
- If we parsed a full command, then `rest.len() == 0` and we drain the whole `v`.
- Otherwise we drain some partial bit of `v`, and we think that shouldn't happen but we handle it anyway.

I think the only reason to _not_ do this would be if the common case (we drain the entire `v`) is much slower than just `v.clear()`. Looking at the impl of `Drain`, I think it is.

I think the way it is is actually optimal - the common case (all of `v` is parsed) is on the fast path with `v.clear()`; the uncommon case (multi-line response) is handled transparently, and we handle the unexpected case of a partial parse but don't expect to ever hit it. I could add a `debug_assert!()` inside the `rest.len() != v.len()` clause, but if we really don't think it's possible then we might as well just put a `panic!()` in there and not bother with `drain()` at all.

Reply:
So, the reason I like a `debug_assert!` in there is that we get a panic to tell us that this _can_ happen, but we don't break "real" code and instead handle that correctly if it turns out that it _is_ possible. But I'm okay going whichever way you want with this.

I think the part I'm missing is why the `rest.len() != v.len()` check is necessary. Why can't it just be a simple `else`? If `rest.is_empty()` then there's no remainder in the input (we parsed it all), and the only alternative is that there _is_ more left in `rest`, which is what `else` would capture. What does the inequality check guard against?

Reply:
Oh, the only reason to have `rest.len() != v.len()` is to avoid going into this branch at all in the multi-line response case where we don't parse anything from the input (because we only need to do anything if `rest != v` so why waste cycles?).

But I think I see what you're getting at here - we can do the `debug_assert` that flags the unexpected behaviour and also does the right thing anyway. I didn't see it until I tried to implement what you were getting at. I pushed a change that I think is what you mean?